opencog/spacetime/octomap/OctoMapValue.cc

```diff
@@ -0,0 +1,69 @@
+#include "OctoMapValue.h"
+
+#include <opencog/spacetime/octomap/atom_types.h>
+#include <opencog/util/exceptions.h>
+#include <opencog/atoms/base/ClassServer.h>
+#include <opencog/atoms/value/ValueFactory.h>
+
+using namespace opencog;
+
+
+void OctoValue::update()
+{
+    //point3d p = (_om->get_locations_of_atom(_octo_atom)).front();
+    //_value = {p.x(), p.y(), p.z()};
+    _om->insert_atom(point3d(_value[0], _value[1], _value[2]), _item);
+}
+
+OctoValue::OctoValue(const HandleSeq& hseq,
+                     std::vector<double> values) : FloatValue(OCTO_VALUE, values)
+{
+
+    if( hseq.size() != 2) {
+        throw InvalidParamException(TRACE_INFO, 
+                "Required 2 arguments. Provided %d.", hseq.size());
+    }
+
+    _item = hseq[0];
+    _octo_atom = hseq[1];
+    _om = OctoMapNodeCast(_octo_atom)->get_map();
+
```

Comment:
and, here, of course, you need to say something like
```
   _om->insert_atom(_item);
```
so that the octomap knows it needs to start tracking the location of `_item`.  That is the whole point of `OctoValue` -- to force the octompa to track the atom, and to ask for the updated location, whenever it is needed.

Reply:
So, you are suggesting OctoValue should only be used as a getter? and whenever **cog-set-value!** is called, all it does is get latest location from the octomap and update the _value vector object?  That makes sense to me.

Reply:
Yes, it is only a "getter". In fact, `atomtypes.script` should probably have `OCTO_VALUE <- STREAM_VALUE`  -- the idea here is that stream values are values that "automatically" change, under the covers, and the atomspace doesn't have to do any work when they change. Examples would be audio or video streams -- so the for example, a stream would store a URL for a webcam or a microphone; it wouldn't actually store the video frames at any given moment; it would just return them, but only if someone asked for them.

See RandomStream.cc as a (really super-simple) example of  stream. It returns a different random number every time you ask for its value.

So, for all streams, the `cog-set-value!` is done only once, to attach the stream to the atom.

Reply:
Oh, and since the constructor does `_om->insert_atom(_item);` the destructor needs to do a `_om->remove_atom(_item);`

Hopefully, it will never happen that there are two different values tracking the same atom in the same octomap. That would be weird and buggy.  So lets just ignore that for now?

Reply:
And part of the idea of using streams in atoms is that the atoms could maybe represent how stream values are hooked together. So maybe there might be some "output stream" that could be a phone or tablet, and to send the video from webcam to the phone, you would attach `(ConceptNode "my webcam")` to `(ConceptNode "that tablet")`  or something like that. Details TBD, no one has ever done this; I haven't thought this through very much.

For the octomap, we would need something like
```
EvaluationLink
     GroundedPredicateNode "scm:is-to-the-left-of"
     ListLink
           Concept "Visible Person 42"
           Concept "visible person 69"
```
and "scm:is-to-the-left-of" would dig out the OctoMapValue for both, compare x,y,z, and return true or false. But it would do this only at the time that the evaluationlink is actually evaluated.

Reply:
And just two more short comments to clarify: 

> `OctoValue` should only be used as a getter? and whenever `cog-set-value!`

You can add setter-methods to the c++ `OctoValue` class. I don't care. It might be an OK way to update the contents of the octomap.  How do you plan to get ROS data from ROS to the octomap?

>  and whenever `cog-set-value!`

Please note that `cog-set-value!` only changes the value on the atom. It does NOT change the value itself! You cant to avoid calling this, as much as possible, because it is a performance killer. Using it requires you to run the constructor for a value, malloc memory, update the KVP database in the atom, and then run the destructor for the old value. This would be an extremely bad idea, if the only goal is to change three floats.

Reply:
the entire point of having values is to be able to change them as frequently as possible right? is there any way in which the cog-set-value! function could be written where changing a value simply comes down to just indexing the vector and changing the value right there? or maybe implementing the virtual function `update()` ? 
also maybe `cog-set-value!` shouldn't be able to create new values at all... like `set!` if the variable isn't already defined, it should give out an error. It only calls the update function of each value type to change the actual value and `cog-new-value` could create new values taking the key pair atoms (because what cog-new-value does is basically the same thing as calling the functions FloatValue StringValue....)

Reply:
on `cog-set-value!` not creating new values(objects) I see that a call to it (for instance on a FloatValue) means a call to `createFloatValue` to attach to the key pair atoms which essentially creates a new FloatValue object (thereby a new vector of floats with the new value) 
Do you think it's possible to use the object's referece to the vector of floats in order to change a specific value from 0.2 to 0.4 without allocating a new memory space for the new FloatValue object? 

I see that the current implementation of values is much much faster than atoms when manipulating them but if there is a way to avoid the above scenario, it would be as fast as it can be.

Reply:
Well, `cog-set-value!`, the scheme code, does have to convert a list of SCM smobs into an instance of a C++ `class Value`, so, yes, it calls the factory to call `createValue` to do this.  After this, it calls the c++ code `Atom::set_value()` which takes a `ValuePtr`, and so that had to come from somewhere.

The SCM smobs are opaque data types managed by guile, they are not floats or doubles or vectors; they're a bucket of bitflags and weird encodings, including "boxes" (read about "boxing" in guile) and other strange stuff. 

The Atomspace Values themselves are designed to be "readonly" (immutable) so as to avoid multi-threading race conditions. If you made them read-write, then you would have to take a lock every time you read them, which would make them less efficient.  My goal here was to write lock-free code. 

I haven't read the following, but I assume it contains valuable explanations: https://softwareengineering.stackexchange.com/questions/171253/does-immutability-entirely-eliminate-the-need-for-locks-in-multi-processor-progr

The general background:  https://en.wikipedia.org/wiki/Immutable_object

Reply:
The OctoValue should not track. Instead, the `update()` function should call `TimeOctoMap` to obtain the current position.  Basically, OctoValue should use zero CPU time, unless someone is actually calling the `get_value()` method.

Reply:
The original comment, stating that ` _om->insert_atom(_item);` still seems to not have been done (or, if its not needed, I don't see why)

Reply:
@misgeatgit yes, the implementation of `update()` now looks correct.

Reply:
Insertion is going to be made by using pointmemory library (https://github.com/opencog/opencog/tree/master/opencog/spacetime/pointmemory). So, the next step will be modifying pointmemory API to return Values instead of Handles. That is what I was thinking. As you said, the other way would be, also doing the insertion using OctoValue. But I am not sure how we can reuse the functionalities provided in pointmemory if we go that way.

Reply:
Yes, that's right, about the insert.  I'd forgotten the pointmemeory API.

Yes, PointMemory needs to be converted to use mostly FloatValues for most things, and maybe OctoValue in one or two places.  That conversion is going to be icky ... it won't always be obvious what to convert, when.   Convert the xyz first, make that a pull request, and then do the rest in a different pull req. That should make things simpler to review and test.